### PR TITLE
LPS-115507

### PR DESCRIPTION
**Pull request by rijema**

### CODE REVIEW

**Comment by rijema** [comment]
ci:test:relevant

**Comment by rodrigopaulino** [comment]
I believe you have a SF failure there, @rijema. Could you please fix it? Thx.

**Comment by rodrigopaulino** [comment]
ci:stop

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java [hunk 1]
```diff
@@ -149,6 +149,14 @@ protected void doProcessAction(
 		String redirectURL = ParamUtil.getString(
 			actionRequest, "redirect", ddmFormInstanceSettings.redirectURL());
 
+		DDMFormSuccessPageSettings ddmFormSuccessPageSettings =
+			ddmForm.getDDMFormSuccessPageSettings();
+
+		if (Validator.isNotNull(redirectURL) ||
```
Comment (rodrigopaulino) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: You don't need to validate the same thing twice. Just include `hideDefaultSuccessMessage(actionRequest);` inside the `if` body and keep the `else`.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java [hunk 2]
```diff
@@ -149,11 +149,15 @@ protected void doProcessAction(
...
 			actionRequest, "redirect", ddmFormInstanceSettings.redirectURL());
 
+		String currentURL = ParamUtil.getString(actionRequest, "currentURL");
+
 		DDMFormSuccessPageSettings ddmFormSuccessPageSettings =
 			ddmForm.getDDMFormSuccessPageSettings();
 
-		if (Validator.isNotNull(redirectURL) ||
+		if ((Validator.isNotNull(redirectURL) &&
+			 !StringUtil.equals(currentURL, redirectURL)) ||
```
Comment (rodrigopaulino) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java:158]: Could you please give more details on why do you need to verify this?

**Review by rodrigopaulino — CHANGES_REQUESTED** [review]

**Comment by rodrigopaulino** [comment]
ci:stop

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 2 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Hi @rodrigopaulino . Please imagine a scenario where the user is creating a Form that he sets a Redirect URL and he doesn't have a success page in his form. This condition allows that when returning to the form, the toast message is not thrown on the screen, whether the form is created with or without a success page.

Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Thanks for explaining, @rijema. What I don't understand is when the currentURL will be equals to the redirectURL. Do you have an example of that?

Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: When we don't set an explicit redirectURL, the code interprets that we're doing a redirection to the same form. In this case, the redirectURL will be the same as the current. Does it make sense to you, @rodrigopaulino?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 1 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: That's not quite what I was thinking. What happens if there's no redirectURL? Would the success page never be verified?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 2 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Does that mean that there's always going to be redirectURL?

Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Wouldn't it be easier to just check if the setting exists?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 1 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: During tests, I didn't see the object _redirectURL_ being empty in different cases I tested. So, I assumed that the success page will always be verified. Both cases reported [here](https://github.com/liferay-forms/liferay-portal/pull/1961#discussion_r982860925) are covered.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 2 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: #`Does that mean that there's always going to be redirectURL?`

Yes. 

#Wouldn't it be easier to just check if the redirectURL setting exists?

I don't think so, since the object will always be filled with an URL, according to the way it was created. I mean, in my testings with different forms I could not found a single use case the value would be empty. But please tell me if I'm being too conclusive in defining this. I can do some more testing with other use cases.

Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Hi, @rijema. Thanks for explaining your thought process. I trust your investigation and judgement. I just want to understand the scenarios in order to come up with the best solution. Now, here's what I'm thinking...
If the redirectURL is never null, why are we checking [this](https://github.com/liferay-forms/liferay-portal/pull/1961/files#diff-d6f8076fb627c946e6e93173fb0dfe74cd724f71a17b99263707b90f5d5e77e4R152) then? Also, I think I wasn't clear before when I said "redirectURL setting". What I meant by that was `ddmFormInstanceSettings.redirectURL()`. Wouldn't it be easier to simply verify if there's a `ddmFormInstanceSettings.redirectURL()` (the one that you type and set inside the form settings modal) to hide the default success message?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java [hunk 3]
```diff
@@ -150,20 +150,24 @@ protected void doProcessAction(
...
+
 				hideDefaultSuccessMessage(actionRequest);
 			}
+
+			portletSession.setAttribute(
+				DDMFormWebKeys.DYNAMIC_DATA_MAPPING_FORM_INSTANCE_ID,
+				formInstanceId);
+			portletSession.setAttribute(DDMFormWebKeys.GROUP_ID, groupId);
+
+			sendRedirect(actionRequest, actionResponse, redirectURL);
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java:170]: Hi @rijema,

I noticed you merged the `if/else`, but do we really need this extra logic in the `if` statement? I understand that, after your tests, you didn't find a use case where the `redirectURL` value was empty, right?

Wouldn't it be the case of deleting the `else` instead of merging with the `if`, since it was a code that wasn't executing before?

What do we really need here to fix the reported issue?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 3 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Hello @carolmariaabb!

This pull request was discussed on the PTR https://issues.liferay.com/browse/PTR-3312, where I would like to update the use cases for successful form submission messages.

First, I tried to update the branch with a clearer commit of my goal and also with some suggestions discussed here. Please see the new unique commit.

Goal:

According to the implementation made, I try to remove the green alert (ToastAlertContainer) from submissions that contain a redirect link and also success pages, based on the PTR discussion.

Answering your questions:

> I noticed you merged the if/else, but do we really need this extra logic in the if statement? I understand that, after your tests, you didn't find a use case where the redirectURL value was empty, right?

That's correct. RedirectURL is never empty.

> Wouldn't it be the case of deleting the else instead of merging with the if, since it was a code that wasn't executing before?

Well, the initial proposal was to take this else as a code block to be evaluated before the if, but after the first iterations in PR, I brought it into the if. That is, the content of the else is within the scope that I created for the solution of this fix, since pages that have a success page could also omit the Green Toast Alert, avoiding duplication of information in the posted content (Alert + SuccessPage).

> What do we really need here to fix the reported issue?

Basically, prevent that when returning to the form, after being redirected to another page, and the notification message from being launched.

Comment (carolmariaabb) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Hi @rijema,

If I understood correctly, after your changes, there is no more green alert when Forms is using the **default** success page and **not** the Redirect URL on Success, right? 

But this scenario was not discussed on the PTR,

> In addition, according to the current implementation, when the setting Redirect URL on Success is configured then the Success Page and the success notification won't be displayed, because Forms understands that the redirected URL will provide the necessary success information to the user who just submitted the form."

The comment above is only mentioning the use case where Redirect URL on Success is configured.

Could you confirm in the PTR the expected behavior when using the default success page?

Also, I believe that some Poshi tests will fail, as we are expecting the green alert to appear when the form is submitted using the default success page.

Thanks!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java [hunk 4]
```diff
@@ -146,25 +146,26 @@ protected void doProcessAction(
 		DDMFormInstanceSettings ddmFormInstanceSettings =
 			ddmFormInstance.getSettingsModel();
 
+		String successRedirectURL = ddmFormInstanceSettings.redirectURL();
+
 		String redirectURL = ParamUtil.getString(
-			actionRequest, "redirect", ddmFormInstanceSettings.redirectURL());
+			actionRequest, "redirect", successRedirectURL);
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java:152]: @rijema Please move this to be right before the `sendRedirect(actionRequest, actionResponse, redirectURL);`, as it is only used there.

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 3 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Of course, @carolmariaabb . Thanks for your entries. I concluded that because the code was understanding the default success page as a redirected URL, I included it in the restriction. I'll ask one more question on the PTR. Thanks in advance!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 4 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Done.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java [hunk 5]
```diff
@@ -146,25 +146,26 @@ protected void doProcessAction(
 		DDMFormInstanceSettings ddmFormInstanceSettings =
 			ddmFormInstance.getSettingsModel();
 
+		String successRedirectURL = ddmFormInstanceSettings.redirectURL();
+
```
[the remark this answers is not in the document]
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: Done.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java (hunk 3 above)
Comment (rijema) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-web/src/main/java/com/liferay/dynamic/data/mapping/form/web/internal/portlet/action/AddFormInstanceRecordMVCActionCommand.java]: According to the last team comment on https://issues.liferay.com/browse/PTR-3312, if success page is defined, success green alert should be on the page. Pull request updated. Regards!

**Review by carolmariaabb — APPROVED** [review]
LGTM!

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by carolmariaabb** [comment]
ci:test:sf

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
Hi @rijema,

The failures don't seem related to your changes. Since CI is taking a long time to run `ci:test:relevant`, feel free to manually forward this to Brian.

Thanks!

**Comment by carolmariaabb** [comment]
ci:stop
